core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -182,6 +182,14 @@ class Partition(val topicPartition: TopicPartition,
   @volatile var leaderReplicaIdOpt: Option[Int] = None
   @volatile var inSyncReplicas: Set[Replica] = Set.empty[Replica]
 
+  // Logs belonging to this partition. Majority of time it will be only one log, but if log directory
+  // is getting changed (as a result of ReplicaAlterLogDirs command), we may have two logs until copy
+  // completes and a switch to new location is performed.
+  // log and futureLog variables defined below are used to capture this
+  @volatile var log: Option[LogInfo] = None
```

Comment:
I think it's true that we would never have a `Partition` object without a corresponding `Log`. It would be nice if we could find a way to represent that so that we could get rid of `localLogOrException`. Potentially we could leave this for a future improvement if this is a lot of work.

Reply:
As discussed, this will be done in a separate JIRA as current commit is pretty big.

Reply:
Hi @hachikuji, I have forced push new changes (as original review lost meaning after moving more code around). Please take a look when you get time.